[Patch 1/2] Feature/client capability response [Test 4.6.2]

CoralMalachi opened the pull request:
### Description
According to Multi-AP Specification, If a Multi-AP Agent receives a
Client Capability Query message, then within one second it shall respond
with a Client Capability Report message.

If the STA specified in the Client Capability Query message is not
associated with any of the BSS operated by the Multi-AP Agent (an error
scenario), the Multi-AP Agent shall set the result code in the Client
Capability Report TLV to 0x01  and include an Error Code TLV.

### Changes
Therefore, the following changes were made:

- Create a Client Capability Report message CMDU.

- Add the Client Info TLV.

- Add a Client Capability Report TLV.

- Check that the Multi-AP Agent has the BSSID that was specified in the Client Info TLV, and that the client is associated with that BSS. If not, this is an error scenario, so set the result code in the Client
Capability Report TLV to 0x01, and include an Error Code TLV with the reason code set to 0x02.

### Testbed status
This branch was tested on testbed while using Qualcomm as CTT Controller, Marvell STA as client and prplMesh as MAUT.
Currently, there is an issue with the Marvell STA, which can not connect to the prplMesh agent.
Since the client was not connected to the agent, the Client Capability Report contained an Error Code TLV as it should.

_Status update: test 4.6.2 **passed** the test on the testbed._

### Next steps

- [x] After solving the issue with the Marvell STA, run the test again and verify that in case of success the Client Capability Report does not contain an Error Code TLV.


- [ ] Client Capability Report TLV  should include the frame body of the most recently received (Re)Association Request frame from this client (if it is an error scenario this field is omitted).
This part will be done in a separate PR.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: update tlvf_conf

Preparative commit.

Add the following TLVs to the `tlvf_conf` YAML file so the TLVF will
auto-generate their source files. Those TLVS will be used in the Client
Capability Report message.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: add TLVs auto-generated files

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: add TLVs to parse method

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller: subscribe to client capability report message

Preparative commit.

According to section 4.6.2 in the EasyMesh Test Plan v1.0 [1],
the MAUT (MultiAP Agent Under Test) sends Client Capability Report
message to the CTT Controller.

To make sure the Client Capability Report message was received by the
controller, The controller subscribes to a message of type :
CLIENT_CAPABILITY_REPORT_MESSAGE.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
controller: add handler to client capability report message

Add handle_cmdu_1905_client_capability_report_message method which prints
"Received CLIENT_CAPABILITY_REPORT_MESSAGE" when the report is received
by the controller.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

tomereli reviewed (approved):
Some comments I'd like your feedback on before merge BUT nothing serious so approving.

arnout commented on common/beerocks/bcl/include/bcl/network/network_utils.h:
```diff
@@ -136,4 +136,11 @@ inline el::base::MessageBuilder &operator<<(el::base::MessageBuilder &log, const
+    auto s1 = beerocks::net::network_utils::mac_to_string(lmac);
+    auto s2 = beerocks::net::network_utils::mac_to_string(rmac);
+    return (s1.compare(s2));
+}
```
This is only needed because you're using a std::set or std::map somewhere, which you shouldn't do.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -2115,6 +2115,16 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
 
+        auto it = m_vaps_clients_map.find(notification_in->params().bssid);
+        if (it == m_vaps_clients_map.end()) {
+            std::set<sMacAddr> clients;
```
std::unordered_set

CoralMalachi replied:
Thanks for pointing it, I'll change it

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1933,6 +1933,12 @@ bool slave_thread::handle_cmdu_ap_manager_message(Socket *sd,
 
+        //update the map to remove this client from its vap
+        auto it = m_vaps_clients_map.find(notification_in->params().bssid);
+        if (it != m_vaps_clients_map.end()) {
```
We might want to throw a warning if the vap is not found...

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4251,7 +4254,67 @@ bool slave_thread::handle_client_capability_query(Socket *sd, ieee1905_1::CmduMe
+            //only if the client exist in one of the BSS operated by the Multi-AP Agent this is not an error scenario
+            error_scenario = 0;
+        }
+    }
```
This is not accurate... If the client is associated with a BSS on another radio in the same agent, both radios will receive the query, but it is not associated with this radio. We shouldn't treat that as an error.

Also, the bssid given in the client info should be treated as just a hint. The spec says "If the STA specified in the Client Capability Query message is not associated with any of the BSS operated by the Multi-AP Agent". It doesn't say "if it is not associated with the BSS given by the client info TLV". A likely scenario is that a client associates and immediately disassociates and moves to another band, so it's possible that the controller sends the query still using the old BSS (which it got from the topology notification).

Bottom line, I think the best approach is to again cache everything on the backhaul_manager and generate the message there.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4251,7 +4254,67 @@ bool slave_thread::handle_client_capability_query(Socket *sd, ieee1905_1::CmduMe
+        LOG(ERROR) << "addClass wfa_map::tlvClientInfo has failed";
+        return false;
+    }
+    client_info_tlv_t->bssid()      = client_info_tlv_r->bssid();
```
bssid may have changed already...

For all we know, stupid controllers may just leave it at 00:00:00:00:00:00.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -972,6 +974,14 @@ bool master_thread::handle_cmdu_1905_steering_completed_message(const std::strin
+    const std::string &src_mac, ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    auto mid = cmdu_rx.getMessageId();
+    LOG(INFO) << "Received CLIENT_CAPABILITY_REPORT_MESSAGE, mid=" << std::hex << int(mid);
```
Please also log the details so we can check it in test_flows.

CoralMalachi replied:
Client Capability Report TLV should include the frame body of the most recently received (Re)Association Request frame from this client (if it is an error scenario this field is omitted).
This part will be done in a separate PR.
So, for now the log message can contain if this is an error scenario or not, and I'll add a TODO comment to add the details later.

arnout reviewed (changes requested):
I unfortunately have some fundamental architectural comments:

> Bottom line, I think the best approach is to again cache everything on the backhaul_manager and generate the message there.

To be verified with @tomereli if he agrees.

tomereli commented:
> I unfortunately have some fundamental architectural comments:
> 
> > Bottom line, I think the best approach is to again cache everything on the backhaul_manager and generate the message there.
> 
> To be verified with @tomereli if he agrees.

I completely agree and I think I also suggested that in my review, didn't want to block merge on this but on second thought I think you are right, we should spend the extra effort and start using the VAPs map for everything - clients, client capabilities, etc.

Commit:
agent: son_slave: remove client capability handling

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
agent: backhaul manager: add client capability handler

When a Multi-AP Agent receives a Client Capability Query message, a
Client Capability Report message is created with the same MID that was
received in the Client Capability Query message.
First, add the Client Info TLV. Then we need to check that this is not an
error scenario :
check that the Multi-AP Agent has the BSSID that was specified in the
Client Info TLV, and that the client is associated with that BSS.
If not, this is an error scenario, so set the result code in the Client
Capability Report TLV to 0x01, and include an Error Code TLV with
the reason code set to 0x02.
In addition, Client Capability Report TLV  should include the frame body
of the most recently received (Re)Association Request frame from this
client (if it is an error scenario this field is omitted).
This part will be done in a separate PR.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
test_flows: update client capability test

Update the test flow to validate the following :

A Client Capability Report message is received by the controller.
If the STA specified in the Client Capability Query message is not
associated with any of the BSS operated by the Multi-AP Agent, it is an
error scenario so the test makes sure that an Error Code TLV was added by
searching the matching log.
If everything is ok,  the test makes sure that the result code in the
Client Capability Report TLV to 0x00 (Success) by searching the matching
log.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -960,6 +963,27 @@ bool master_thread::handle_cmdu_1905_steering_completed_message(const std::strin
+            ? "SUCCESS"
+            : "FAILURE";
+
+    //TODO: log the details so it can be checked in the test_flows
```
:1st_place_medal:

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1814,6 +1818,76 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+    }
+
+    //Check if it is an error scenario - if the STA specified in the Client Capability Query message is not associated
+    //with any of the BSS operated by the Multi-AP Agent (an error scenario)
```
It may be useful to explicitly mention that even though client_info_tlv_r does contain a BSSID, the specification says that we should answer if the client is associated with *any* BSS on this agent.

CoralMalachi replied:
Agree.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1814,6 +1818,76 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+
+    //Check if it is an error scenario - if the STA specified in the Client Capability Query message is not associated
+    //with any of the BSS operated by the Multi-AP Agent (an error scenario)
+    bool error_scenario = 1;
```
Use `true`/`false` for booleans, not 1 and 0.

Also, instead of `error_scenario`, I think it makes more sense to call it e.g. `associated_client_found` (and inverse the logic, obviously).

(nitpick)

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1814,6 +1818,76 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
+    if (!error_scenario) {
+        client_capability_report_tlv->result_code() = wfa_map::tlvClientCapabilityReport::SUCCESS;
+        LOG(DEBUG) << "Result Code: SUCCESS";
+        //TODO: Add frame body of the most recently received (Re)Association Request frame from this client
```
So, this...

We have two options here. Either we send a message to the slave that forwards it to the ap_manager that converts it to a bwl call and generates a response message to the slave that gets forwarded to the backhaul manager that generates the response message (with the same MID). Or, we add the ReAssociation Request frame to the CLIENT_ASSOCIATED_NOTIFICATION and cache it in the backhaul manager. The latter would also require us to resend the notification every time a new reassociation request is received by the AP.

If we choose the first option, we have to take into account a possible race condition: by the time the message reaches hostapd, the client may already have left the BSS.

@vitalybu @morantr @tomereli your opinions?

Note that for this PR it doesn't matter much. If it is cached, we'll just need to update the loop in the beginning that finds the associated client to not just set a flag, but to set (a pointer to) the association request frame.

tomereli replied:
I prefer option (2) - caching. I think it is cleaner.
However, the race condition may occur in both options since the client may have left also by the time the message with the updated ReAssociation frame is received in the BH manager

CoralMalachi replied:
I also prefer the second option.

arnout commented on tests/test_flows.sh:
```diff
@@ -211,15 +211,31 @@ test_channel_selection() {
     check_error=0
-    send_CAPI_1905 ${GATEWAY} $mac_agent1 0x8009 "tlv_type,0x90,tlv_length,\
-0x000C,tlv_value,{$mac_agent1_wlan0 0x000000110022}"
+    check send_CAPI_command ${GATEWAY} "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8009,tlv_type,0x90,tlv_length,\
```
Please use send_CAPI_1905 function.

arnout commented on tests/test_flows.sh:
```diff
@@ -211,15 +211,31 @@ test_channel_selection() {
+    check_log ${REPEATER1} agent "CLIENT_CAPABILITY_QUERY_MESSAGE"
+
+    dbg "Confirming that Client Capability Report message contains an Error Code TLV"
+    check_log ${REPEATER1} agent "Result Code: FAILURE"
```
It would be better to check this on the controller instead of the agent. The commit that adds the handler on the controller could very easily parse out the result code and report it.

Also, to make sure the correct log message is checked, add the client MAC and BSSID to the output.

arnout commented on tests/test_flows.sh:
```diff
@@ -211,15 +211,31 @@ test_channel_selection() {
+    dbg "Connect dummy STA to wlan0"
+    send_bwl_event ${REPEATER1} wlan0 "EVENT AP-STA-CONNECTED ${sta_mac}"
+
+    check send_CAPI_command ${GATEWAY} "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8009,tlv_type,0x90,tlv_length,\
```
Again, send_CAPI_1905 please.

arnout commented on tests/test_flows.sh:
```diff
@@ -211,15 +211,31 @@ test_channel_selection() {
+0x000C,tlv_value,{$mac_agent1_wlan0 0x000000110022}" $redirect
+    sleep 1
+
+    check_log ${REPEATER1} agent "Result Code: SUCCESS"
```
Same here, better to check it on the controller.

arnout commented on tests/test_flows.sh:
```diff
@@ -211,15 +211,31 @@ test_channel_selection() {
-    send_CAPI_1905 ${GATEWAY} $mac_agent1 0x8009 "tlv_type,0x90,tlv_length,\
-0x000C,tlv_value,{$mac_agent1_wlan0 0x000000110022}"
+    check send_CAPI_command ${GATEWAY} "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8009,tlv_type,0x90,tlv_length,\
+0x000C,tlv_value,{$mac_agent1_wlan0 0x000000110022}" $redirect
```
Please use a *different* sta_mac here than the one for the success scenario. This helps to make sure we parse out the correct log message. Cfr. #822

arnout commented on tests/test_flows.sh:
```diff
@@ -211,15 +211,31 @@ test_channel_selection() {
+    send_bwl_event ${REPEATER1} wlan0 "EVENT AP-STA-CONNECTED ${sta_mac}"
+
+    check send_CAPI_command ${GATEWAY} "DEV_SEND_1905,DestALid,$mac_agent1,MessageTypeValue,0x8009,tlv_type,0x90,tlv_length,\
+0x000C,tlv_value,{$mac_agent1_wlan0 0x000000110022}" $redirect
```
You can actually use `${sta_mac}` here - our CAPI parser accepts it.

arnout reviewed (approved):
*Very* good splitting in commits and good commit messages. :1st_place_medal: 

Of course I have a few nitpicks, but we can go through with merging this.

There is one major question remaining about how it should be implemented.
